core/src/main/java/hudson/PluginManager.java

```diff
@@ -778,45 +945,115 @@ public HttpResponse doUpdateSources(StaplerRequest req) throws IOException {
+        
+        for (int i = 0; i < pluginListJSON.size(); i++) {
+            plugins.add(pluginListJSON.getString(i));
+        }
+
+        UUID correlationId = UUID.randomUUID();
+        try {
+            boolean dynamicLoad = request.getBoolean("dynamicLoad");
+            install(plugins, dynamicLoad, correlationId);
+
+            JSONObject responseData = new JSONObject();
+            responseData.put("correlationId", correlationId.toString());
+
+            return new JSONObjectResponse(responseData);
+        } catch (Exception e) {
+            return new JSONObjectResponse().error(e.getMessage());
+        }
+    }
+
+    /**
+     * Performs the installation of the plugins.
+     * @param plugins The collection of plugins to install.
+     * @param dynamicLoad If true, the plugin will be dynamically loaded into this Jenkins. If false, 
+     *                    the plugin will only take effect after the reboot.
+     *                    See {@link UpdateCenter#isRestartRequiredForCompletion()}
+     * @return The install job list.                   
+     * @since FIXME                    
+     */
+    public List<Future<UpdateCenter.UpdateCenterJob>> install(@Nonnull Collection<String> plugins, boolean dynamicLoad) {
+        return install(plugins, dynamicLoad, null);
+    }
+    
+    private List<Future<UpdateCenter.UpdateCenterJob>> install(@Nonnull Collection<String> plugins, boolean dynamicLoad, @CheckForNull UUID correlationId) {
+        List<Future<UpdateCenter.UpdateCenterJob>> installJobs = new ArrayList<>();
+        
+        for (String n : plugins) {
+            // JENKINS-22080 plugin names can contain '.' as could (according to rumour) update sites
+            int index = n.indexOf('.');
+            UpdateSite.Plugin p = null;
+            
+            if (index == -1) {
+                p = getPlugin(n, UpdateCenter.ID_DEFAULT);
+            } else {
                 while (index != -1) {
                     if (index + 1 >= n.length()) {
                         break;
                     }
                     String pluginName = n.substring(0, index);
                     String siteName = n.substring(index + 1);
-                    UpdateSite updateSite = Jenkins.getInstance().getUpdateCenter().getById(siteName);
-                    if (updateSite == null) {
-                        throw new Failure("No such update center: " + siteName);
-                    } else {
-                        UpdateSite.Plugin plugin = updateSite.getPlugin(pluginName);
-                        if (plugin != null) {
-                            if (p != null) {
-                                throw new Failure("Ambiguous plugin: " + n);
-                            }
-                            p = plugin;
+                    UpdateSite.Plugin plugin = getPlugin(pluginName, siteName);
+                    // TODO: Someone that understands what the following logic is about, please add a comment.
```

Comment:
I guess it means we do not know which update site it is supposed to be loaded from, when >1 are configured.